graphqlmetrics/core/telemetry_test.go

```diff
@@ -0,0 +1,283 @@
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"os/signal"
+	"strings"
+	"syscall"
+	"testing"
+	"time"
+
+	"connectrpc.com/connect"
+	"github.com/phayes/freeport"
+	"github.com/prometheus/client_golang/prometheus"
+	io_prometheus_client "github.com/prometheus/client_model/go"
+	"github.com/stretchr/testify/assert"
+	graphqlmetricsv1 "github.com/wundergraph/cosmo/graphqlmetrics/gen/proto/wg/cosmo/graphqlmetrics/v1"
+	"github.com/wundergraph/cosmo/graphqlmetrics/gen/proto/wg/cosmo/graphqlmetrics/v1/graphqlmetricsv1connect"
+	"github.com/wundergraph/cosmo/graphqlmetrics/pkg/telemetry"
+	"github.com/wundergraph/cosmo/graphqlmetrics/test"
+	"go.uber.org/zap"
+	brotli "go.withmatt.com/connect-brotli"
+)
+
+const (
+	// this token was generated using the local secret and is therefore already corrupted
+	bearerToken = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJmZWRlcmF0ZWRHcmFwaElEIjoiZmVkMTIzIiwib3JnYW5pemF0aW9uSUQiOiJvcmcxMjMiLCJpYXQiOjE3MjIyNTU5NTR9.8mxFEDqmzmmhPVfKedzTuUUM4VxvPnsPP5N3_8fnecY"
+)
+
+func newServerCtx() (context.Context, context.CancelFunc) {
+	return signal.NotifyContext(context.Background(), os.Interrupt,
+		syscall.SIGHUP,  // process is detached from terminal
+		syscall.SIGTERM, // default for kill
+		syscall.SIGKILL,
+		syscall.SIGQUIT, // ctrl + \
+		syscall.SIGINT,  // ctrl+c
+	)
+}
+
+func TestExposingPrometheusMetrics(t *testing.T) {
+	if os.Getenv("INT_TESTS") != "true" {
+		t.Skip("Skipping integration tests")
+	}
+
+	freePort, err := freeport.GetFreePort()
+	assert.Nil(t, err)
+	prometheusListenAddr := fmt.Sprintf("0.0.0.0:%v", freePort)
+
+	type tc struct {
+		name       string
+		prom       telemetry.PrometheusConfig
+		statusCode int
+	}
+
+	tests := []tc{
+		{
+			name: "enabled",
+			prom: telemetry.PrometheusConfig{
+				Enabled:      true,
+				ListenAddr:   prometheusListenAddr,
+				Path:         "/metrics",
+				TestRegistry: prometheus.NewRegistry(),
+			},
+			statusCode: 200,
+		},
+		{
+			name: "disabled",
+			prom: telemetry.PrometheusConfig{
+				Enabled:      false,
+				ListenAddr:   prometheusListenAddr,
+				Path:         "/metrics",
+				TestRegistry: prometheus.NewRegistry(),
+			},
+			statusCode: -1,
+		},
+	}
+
+	db := test.GetTestDatabase(t)
+	msvc := NewMetricsService(zap.NewNop(), db)
+	ctx, stop := newServerCtx()
+	defer stop()
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			svr := NewServer(ctx, msvc,
+				WithListenAddr("0.0.0.0:0"),
+				WithMetrics(&telemetry.Config{
+					Prometheus: tt.prom,
+				}))
+
+			go func() {
+				if err := svr.Start(stop); err != nil && !errors.Is(err, http.ErrServerClosed) {
+					t.Logf("failed starting main server")
+				}
+			}()
+
+			defer func() {
+				err := svr.Shutdown(ctx)
+				if err != nil {
+					t.Fatalf("Failed to shut down server: %v", err)
+				}
+			}()
+			time.Sleep(2 * time.Second)
+
+			endpoint := fmt.Sprintf("http://%s%s", tt.prom.ListenAddr, tt.prom.Path)
+			resp, err := http.Get(endpoint)
+
```

Comment:
Calling it earlier here is not possible as resp could be nil.
I could change the test and split it up in two tests. 

The intention here was to express the two possibilities, either there will be a response, or it will be nil as there was no connection possible.

Reply:
You're right. Let me rephrase.

```go
if err != nil {
// err handling
}
defer resp.Body.Close()

// response handling
```

Reply:
The test case does not allow this as in once case resp can be nil and the error branch is a valid case here. 

"prometheus was not enabled and no response should have been able to be made and is therefore nil". 

I get your point, I moved the defer to the earliest point, where I know, there is something to close. I will let that sink in and think of splitting it up in two tests.

Reply:
Ok